src/v/kafka/protocol/schemata/flex_api_version_generator.py

```diff
@@ -0,0 +1,174 @@
+            return min, max
+
+
+def parse_schema(schema):
+    s = io.StringIO()
+    with open(schema, 'r') as f:
+        for line in f.readlines():
+            line = re.sub("\/\/.*", "", line)
+            if line.strip():
+                s.write(line)
+
+    return json.loads(s.getvalue())
+
+
+def parse_version(f):
+    assert pathlib.Path(f).suffix == '.json'
+    schema = parse_schema(f)
+    assert schema is not None
+    api_key = schema["apiKey"]
+    assert api_key is not None
+    flex = schema['flexibleVersions']
+    assert flex is not None
+    if flex == "none":
+        flex = never_flexible
+    else:
+        r = VersionRange(flex)
+        flex = r.min
+    return api_key, flex
+
+
+def write_versions(flex_version_map, outfile):
+    """
+    For better performance generate code as a flat static array
+    """
+    assert len(flex_version_map) > 0
+    max_api_key = -1
+    for api_key, api_version in flex_version_map.items():
+        max_api_key = max(max_api_key, api_key)
+    arr_mapping = [invalid_api for _ in range(0, max_api_key + 1)]
+    for api_key, api_version in flex_version_map.items():
+        arr_mapping[api_key] = api_version
+
+    cpp_template = """#pragma once
+#include <array>
+#include "kafka/types.h"
+
+// autogenerated file - do not manually edit
+//
+// The following std::array contains a mapping between each kafka api request
+// where the api key is the index into this array which maps to the corresponding
+// first version for which flex support was introduced for that request.
+
+namespace kafka {
+class flex_versions {
+public:
+    static constexpr bool is_flexible_request(api_key key, api_version version) {
+        /// If bounds checking is desired call is_api_in_schema(key) beforehand
+        const api_version first_flex_version = _flex_mapping[key()];
+        return (version >= first_flex_version) && (first_flex_version != never_flexible);
+    }
+
+    static constexpr bool is_api_in_schema(api_key key) noexcept {
+        constexpr auto max_version = _flex_mapping.max_size() - 1;
+        if (key() < 0 || static_cast<size_t>(key()) > max_version) {
+            return false;
+        }
+        const api_version first_flex_version = _flex_mapping.at(key());
+        return first_flex_version != invalid_api;
+    }
+
+private:
+    /// The only request that is never flexible is sasl_handshake_request - 17
+    /// Older versions of our schema may also contain values of 'none' that map to {{ never_flexible }}
+    static constexpr inline api_version never_flexible = api_version({{ never_flexible }});
+
+    /// Not every value from 0 -> max_api_key is a valid request, non-supported
+    /// requests will map to a value of api_key({{ invalid_api }})
+    static constexpr inline api_version invalid_api = api_version({{ invalid_api }});
+
+    static constexpr std::array<api_version, {{ length }}> _flex_mapping{
+{%- for value in flex_version_arr %}
+      api_version({{ value }}),
```

Comment:
this all looks good. i think you might be able to eliminate this separate code generator by making a small maybe < 5 line code modification to the existing generator, and then using the same trick that the api_versions handler does to report supported api version levels.

what i have in mind would look something like:

1. in the existing code generator, add a `static constexpr api_version min_flexible_version = ...` field to each root structure that is already generated today by the generator. For example:

```c++
struct fetch_request_data {
    static auto constexpr min_flexible_version = api_version(7);
...
};
```

2. then in the `api_versions.cc` use the same folding trick over all of the types:

https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/server/handlers/api_versions.cc#L60-L74

and wrap the computed structures in an api similar to how `std::vector<api_versions_response_key> get_supported_apis()` wraps the computed fold.

---

either way, though. this generator looks good.